### PR TITLE
Fix cached promised values not resolving properly in some platforms.  #874 #878 #887

### LINKED ISSUE #882: Commit message lost when git returns error code
When committing something using ungit, and the commit fails in some way with git returning a non-zero error code, the commit message (title and body) entered into ungit is removed.  

This is particularly frustrating when ungit is used with pre-commit hooks.  These hooks modify code and use the non-zero git return code to allow you to inspect and re-commit the automatic changes.  

In an older version of ungit (not sure which, I could bisect), when git / pre-commit hooks returned non-zero code, the commit message was preserved in the ungit UI.  This allowed me to simply press the commit button and see if the pre-commit hooks did anything and press commit again if they did.  

In the current version, if the pre-commit hooks fires, I have to retype my commit message.  This is a bit cumbersome, and I have worked around it at the moment by copying my message before committing in case the hook fires and I have to paste in the text again.

### PR DESCRIPTION
**Pull request by codingtwinky**
* add readme for #887
* fix rare edge case where remote ref node is gone on reset operation
* rescursively resolve all promises before caching them.  #878 #874
* Retain commit messages when commit fails fix #882

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -99,6 +99,7 @@ Known issues
 * If you're running MacOSX Mavericks and Ungit crashes after a few seconds; try updating npm and node. See [#259](https://github.com/FredrikNoren/ungit/issues/259) and [#249](https://github.com/FredrikNoren/ungit/issues/249) for details.
 * Ubuntu users may have trouble installing because the node executable is named differently on Ubuntu, see [#401](https://github.com/FredrikNoren/ungit/issues/401) for details.
 * Debian Wheezy's supported git and nodejs packages are too old, therefore download newest [git](https://github.com/git/git/releases) and [nodejs](https://nodejs.org/download/) tarballs and [build from source](http://www.control-escape.com/linux/lx-swinstall-tar.html).
+* Adblocker may block Ungit! Some ad blockers, such as uBlock, doesn't like localhost api calls and assumes that it is a cross domain attack.  Please whitelist `localhost:{ungit port number}`. [#887](https://github.com/FredrikNoren/ungit/issues/887)
```
Comment (dregad) [inline README.md]: Suggested changes:
- rewording:`Some ad blockers, such as [Adblock plus](https://adblockplus.org) and [uBlock](https://www.ublock.org/), don't like localhost api calls and assume that [...]`
- add reference to `[#892](https://github.com/FredrikNoren/ungit/issues/892)`

Also, note that the host name it is not necessarily *localhost* (in my case, running ungit on Cloud9 workspace, I have a *workspace-name.c9users.io* hostname), so your hint could be misleading.

File: source/utils/cache.js [hunk 2]
```diff
@@ -24,6 +23,21 @@ cache.resolveFunc = (key) => {
...
+ * @description In Linux, or certain settings, it seems that cached promises
+ *   are not able to resolved and we need to cache raw result of promieses.
+ * @param {prom} - raw value or promise to be returned or resolved
+ * @param {promise} - a promise where next "then" will result in raw value.
+ */
+const getHardValue = (prom) => {
+  if (prom.then) {
+    return prom.then(getHardValue);
+  } else {
+    return Bluebird.resolve(prom);
```
Comment (campersau) [inline source/utils/cache.js]: Strange... normally [`.resolve`](http://bluebirdjs.com/docs/api/promise.resolve.html) should already do this:
> If value is a thenable (Promise-like object, like those returned by jQuery's $.ajax), returns a trusted Promise that assimilates the state of the thenable.

Comment (codingtwinky) [inline source/utils/cache.js]: yeah this is a terrifying find and i need more confirmation but behavior was replicate-able in ubuntu.

File: README.md [hunk 3]
```diff
@@ -99,6 +99,7 @@ Known issues
 * If you're running MacOSX Mavericks and Ungit crashes after a few seconds; try updating npm and node. See [#259](https://github.com/FredrikNoren/ungit/issues/259) and [#249](https://github.com/FredrikNoren/ungit/issues/249) for details.
 * Ubuntu users may have trouble installing because the node executable is named differently on Ubuntu, see [#401](https://github.com/FredrikNoren/ungit/issues/401) for details.
 * Debian Wheezy's supported git and nodejs packages are too old, therefore download newest [git](https://github.com/git/git/releases) and [nodejs](https://nodejs.org/download/) tarballs and [build from source](http://www.control-escape.com/linux/lx-swinstall-tar.html).
+* Adblocker may block Ungit! Some ad blockers, such as [Adblock plus](https://adblockplus.org) and [uBlock](https://www.ublock.org/), doesn't like localhost api calls and assumes that it is a cross domain attack.  Please whitelist `{localhost|127.0.0.1|$UngitURL}:{ungit port number}`. [#887](https://github.com/FredrikNoren/ungit/issues/887) [#892](https://github.com/FredrikNoren/ungit/issues/892)
```
Comment (dregad) [inline README.md]: @codingtwinky thanks for updating the README. If I may pick nits, you missed one little bit `doesn't` --> `don't`.

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,9 @@ Use the following format for additions: ` - VERSION: [feature/patch (if applicab
     - Fix cli arguments [#871](https://github.com/FredrikNoren/ungit/pull/871)
     - Stop if ~/.ungitrc contains syntax error
     - Removed official support ini format of ~/.ungitrc, because internal API supports only JSON
+    - Retain commit messages when commit fails [#882](https://github.com/FredrikNoren/ungit/pull/882)
```
Comment (FredrikNoren) [inline CHANGELOG.md]: 1.1.11 is already out right? Think you ment to cut a new version here

**Pull request merged**